1347: Add script to stress test CodaLab backend

teetone opened the pull request:
Resolves #1347 

Usage:
![Screen Shot 2019-12-03 at 1 02 19 AM](https://user-images.githubusercontent.com/16793796/70035893-93e10080-1568-11ea-8a0b-ef4689f5c16c.png)

Commit:
1347: stress test pass 1

Commit:
1347: added ability to define a configurable run for the stress test runner

epicfaace commented:
Can you make sure stress_test.py is copied to the docker container as in https://github.com/codalab/codalab-worksheets/pull/1788/files#diff-4a7fdf57d68d6fd4470c72984e1d15fe ? That way, it will be easy to run it on the dev server since it will already be there.

percyliang reviewed (commented):
Can you comment on what happens when you run the script on dev?  What does the output look like?

teetone commented:
> Can you make sure stress_test.py is copied to the docker container as in https://github.com/codalab/codalab-worksheets/pull/1788/files#diff-4a7fdf57d68d6fd4470c72984e1d15fe ? That way, it will be easy to run it on the dev server since it will already be there.

@epicfaace: I see what you mean now. It was a bit of a hassle to get everything set up. Thanks again for running the tests for me before. I will make the change to `Dockerfile.server` soon.

Commit:
1347: copy over stress_test.py to Docker container

teetone committed:
1347: lowered some of the counts for the HeavyStressTestRunner

teetone committed:
1347: use a process pool for the parallel runs

teetone committed:
Merge branch '1347-stress-test' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone commented:
> Can you comment on what happens when you run the script on dev? What does the output look like?

@percyliang : I still haven't had a successful run with full, heavy stress tests on dev. It either crashes dev or the heartbeats fail.

However, I did a handful of light and custom runs, which were successful. Here is a small subset of how the output will look like:
[light_output.log](https://github.com/codalab/codalab-worksheets/files/3936685/light_output.log)

teetone committed:
1347: fix MB to bytes conversion

percyliang reviewed (commented):
Where do you clean up any worksheet/bundle created by stress tests?  Should behave properly when there are failures...

teetone committed:
1347: added clean up logic

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone commented:
@percyliang  The clean up happens at the end of the stress testing. I currently keep track of all bundles and worksheets and just delete them at the end.

percyliang reviewed (changes requested):
1) Stress tests can most certainly fail and we might want to cancel things in the middle, so it's important to be able to be able to cleanup easily even if that happens, whereas right now, if the script dies halfway through, then we can't cleanup.  I'd suggest tagging any bundle or worksheet that's created by this script with something like `stress-test`, and then cleanup can be something like `cl search tag=stress-test .limit=500 | xargs cl`.  We should print out instructions on how to clean up.  Finally, we should cleanup after each test.

2) I still think the light/heavy config is too verbose and can be simplified.  First, I don't think there's any point in having 3 default configs (light, heavy, and command-line args).  Second, I don't think the configs should be child classes of `StressTestRunner` - they should be simply data.  So I'd just define the command-line arguments as you had them before with the light defaults.  If `--heavy` is set, then just simply set the relevant command-line arguments.  Then pass the `args` object into the `StressTestRunner`.

teetone commented:
> 1. Stress tests can most certainly fail and we might want to cancel things in the middle, so it's important to be able to be able to cleanup easily even if that happens, whereas right now, if the script dies halfway through, then we can't cleanup.  I'd suggest tagging any bundle or worksheet that's created by this script with something like `stress-test`, and then cleanup can be something like `cl search tag=stress-test .limit=500 | xargs cl`.  We should print out instructions on how to clean up.  Finally, we should cleanup after each test.
> 2. I still think the light/heavy config is too verbose and can be simplified.  First, I don't think there's any point in having 3 default configs (light, heavy, and command-line args).  Second, I don't think the configs should be child classes of `StressTestRunner` - they should be simply data.  So I'd just define the command-line arguments as you had them before with the light defaults.  If `--heavy` is set, then just simply set the relevant command-line arguments.  Then pass the `args` object into the `StressTestRunner`.

@percyliang  I am thinking of leaving the current clean up logic that will clean up after the stress tests on a happy path and also adding an additional parameter (something like `--clean-up-only`) that gives the option to just clean up the tagged bundles and worksheets.

One follow up question I had is do we have the capability to search for bundles by their tags i.e. `cl search tag=stress-test`?  `cl wsearch tag=stress-test` works, but I'm not sure if the `tag` option for bundle searching is currently supported or not.

percyliang commented:
The cleanup logic in both cases should exercise the same code path for consistency.

As for searching bundles by tags, it should work (if not, we should make it work).  Note that the search has a limit, so we have to put a loop around it.

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone committed:
1347: added clean up logic for tagged entities

teetone commented:
@percyliang It turns out when searching for bundles, the metadata field name has to be used, so I had to use `tags` for bundles and `tag` for worksheets.

I removed the child classes and cleaned up the configuration logic, so if `heavy` is set to true, I override the default, light configuration. I also added an additional option to just do the cleanup only, which uses the same code path as the post-stress tests cleanup.

teetone committed:
1347: get rid of accessors

teetone committed:
1347: fix styling

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone committed:
1347: use namespace object directly

percyliang reviewed (approved):
This is much better!  Can you also attach as a PR comment a transcript of this script running against some server?  It'd be nice to know what the expected output is and how longs take to run.

teetone committed:
1347: wrap cleanup logic in loops

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone committed:
1347: expose cleanup function

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone committed:
1347: added bypass_wait option

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1347-stress-test

teetone committed:
1347: add additional argument to specify disk write size
